// react-bits/lib/injection/injectors/inject-react-native-web.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
exports.__esModule = true;
var ReactBits_1 = require("../../ReactBits");
var react_native_web_1 = require("react-native-web");
var Touchable_1 = require("../../modules/Touchable");
var registry_1 = require("react-native-web/dist/apis/StyleSheet/registry");
var emptyObject = {};
var resolve = function (style) {
    return registry_1["default"].resolve(style) || emptyObject;
};
ReactBits_1.ReactBits.inject({
    View: react_native_web_1.View,
    Text: react_native_web_1.Text,
    Image: react_native_web_1.Image,
    Easing: react_native_web_1.Easing,
    Animated: react_native_web_1.Animated,
    Dimensions: react_native_web_1.Dimensions,
    StyleSheet: __assign({}, react_native_web_1.StyleSheet, { resolve: resolve }),
    Platform: {
        OS: react_native_web_1.Platform.OS,
        Version: react_native_web_1.Platform.Version
    },
    Touchable: Touchable_1.createTouchable(react_native_web_1.Animated, react_native_web_1.StyleSheet, react_native_web_1.Platform, react_native_web_1.Touchable.Mixin)
});
