// react-bits/lib/injection/injectors/inject-react-sketchapp.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
exports.__esModule = true;
var ReactBits_1 = require("../../ReactBits");
var Animated = require("animated");
var Easing = require("animated/lib/Easing");
var react_sketchapp_1 = require("react-sketchapp");
var Touchable_1 = require("../../modules/Touchable");
var TouchableMixin = {
    componentWillUnmount: function () { },
    touchableGetInitialState: function () {
        return {
            touchable: {
                touchState: undefined,
                responderID: null
            }
        };
    },
    touchableHandleResponderTerminationRequest: function () { return false; },
    touchableHandleStartShouldSetResponder: function () { return false; },
    touchableLongPressCancelsPress: function () { return true; },
    touchableHandleResponderGrant: function () { },
    touchableHandleResponderRelease: function () { },
    touchableHandleResponderTerminate: function () { },
    touchableHandleResponderMove: function () { }
};
Animated.inject.FlattenStyle(react_sketchapp_1.StyleSheet.flatten);
ReactBits_1.ReactBits.inject({
    StyleSheet: react_sketchapp_1.StyleSheet,
    View: react_sketchapp_1.View,
    Text: react_sketchapp_1.Text,
    Image: react_sketchapp_1.Image,
    Easing: Easing,
    Animated: __assign({}, Animated, { View: Animated.createAnimatedComponent(react_sketchapp_1.View), Text: Animated.createAnimatedComponent(react_sketchapp_1.Text), Image: Animated.createAnimatedComponent(react_sketchapp_1.Image) }),
    Platform: {
        OS: 'sketch',
        Version: 1
    }
});
ReactBits_1.ReactBits.inject({
    Touchable: Touchable_1.createTouchable(Animated, react_sketchapp_1.StyleSheet, ReactBits_1.ReactBits.Platform, TouchableMixin)
});
